src/temp_sensor.rs

```diff
@@ -0,0 +1,149 @@
+/// Rust translation of `temperature_sensor_clk_src_t`
+pub enum TemperatureSensorClockSource {
+    Default,
+    XTAL,
+    RcFast,
+}
+
+#[cfg(not(any(esp32)))]
+impl From<TemperatureSensorClockSource> for temperature_sensor_clk_src_t {
+    fn from(value: TemperatureSensorClockSource) -> Self {
+        match value {
+            TemperatureSensorClockSource::XTAL => {
+                soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_XTAL
+            }
+            TemperatureSensorClockSource::RcFast => {
+                soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_RC_FAST
+            }
+            TemperatureSensorClockSource::Default => {
+                soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_DEFAULT
+            }
+        }
+    }
+}
+
+#[cfg(not(any(esp32)))]
+impl From<temperature_sensor_clk_src_t> for TemperatureSensorClockSource {
+    fn from(value: temperature_sensor_clk_src_t) -> Self {
+        match value {
+            #[allow(non_upper_case_globals)]
+            soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_XTAL => Self::XTAL,
+            #[allow(non_upper_case_globals)]
+            soc_periph_temperature_sensor_clk_src_t_TEMPERATURE_SENSOR_CLK_SRC_RC_FAST => {
+                Self::RcFast
+            }
+            // TODO: Perhaps the default value should be mapped explicitly
+            // and all other (u32) values should cause a failure
+            _ => Self::Default,
+        }
+    }
+}
+
+// -- TemperatureSensorConfig --
+
+#[cfg(not(any(esp32)))]
+#[derive(Copy, Clone)]
+/// Rust wrapper for `temperature_sensor_config_t`
+pub struct TemperatureSensorConfig {
+    // TODO: check int size
+    pub range_min: i32,
+    pub range_max: i32,
+    pub clk_src: TemperatureSensorClockSource,
+}
+
+#[cfg(not(any(esp32)))]
+impl From<temperature_sensor_config_t> for TemperatureSensorConfig {
+    fn from(value: temperature_sensor_config_t) -> Self {
+        TemperatureSensorConfig {
+            range_min: value.range_min,
+            range_max: value.range_max,
+            clk_src: value.clk_src.into(),
+        }
+    }
+}
+
+#[cfg(not(any(esp32)))]
+impl Into<temperature_sensor_config_t> for TemperatureSensorConfig {
+    fn into(self) -> temperature_sensor_config_t {
+        temperature_sensor_config_t {
+            clk_src: self.clk_src.into(),
+            range_max: self.range_max,
+            range_min: self.range_min,
+        }
+    }
+}
+
+#[cfg(not(any(esp32)))]
+impl Default for TemperatureSensorConfig {
+    fn default() -> Self {
+        TemperatureSensorConfig {
+            range_min: -10,
+            range_max: 80,
+            clk_src: TemperatureSensorClockSource::Default,
+        }
+    }
+}
+
+// -- TemperatureSensorDriver --
+
+#[cfg(not(any(esp32)))]
+pub struct TemperatureSensorDriver {
+    ptr: temperature_sensor_handle_t,
+}
+
+#[cfg(not(any(esp32)))]
+impl TemperatureSensorDriver {
```

Comment:
This is the first driver that does not take any peripheral.

So here's the question: what happens if I try to instantiate the driver twice?

Reply:
from the [documentation](https://docs.espressif.com/projects/esp-idf/en/stable/esp32c3/api-reference/peripherals/temp_sensor.html#thread-safety):
```
In the temperature sensor driver, we do not add any protection to ensure the thread safety, because typically this driver is only supposed to be used in one task. If you have to use this driver in different tasks, please add extra locks to protect it.
```

Reply:
What I'm asking for does not have anything to do with thread (un)safety. 
Thread unsafety is correctly modeled already by:
- the driver taking `&mut` refs everywhere
- the driver containing a raw pointer (`ptr`) which makes it `!Sync`

(Before I forget: the driver however needs an `unsafe impl Send for TemperatureSensorDriver {}` because currently it is also `!Send` due to the above mentioned `ptr`. However, there is nothing that stops the driver from being `Send` so we should mark it as such.)

What I'm talking about is - consider this:
```rust
let driver1 = TemperatureSensorDriver::new()?;
let driver2 = TemperatureSensorDriver::new()?; // Will likely fail with a runtime error because the C driver is already installed

// there is a deeper problem here however - what if you try to call `TemperatureSensorDriver::new` 
// from TWO separate threads, and thus the two calls race each other? 
// If the C call `temperature_sensor_install` is itself thread-safe, that would work and one of these calls will fail, 
// while the other will succeed; however if it is not thread-safe, you might end up with memory corruption or even worse things
```

The fact the second driver instantiation fails is OK because it does not make sense to have multiple drivers for a **single** peripheral. However, `esp-idf-hal` goes one step further than that: **every driver takes a peripheral**. This way - even if you wanted to instantiate the driver twice, you cannot. At compile time. Including from multiple threads, as only one of the thread can "own" the peripheral at any point in time. I suggest (again) to take a look at the other drivers as to how to model a peripheral.

Reply:
Btw, and if you are struggling with the "peripheral" thing:
- You only need to copy-paste [this line](https://github.com/esp-rs/esp-idf-hal/blob/b9baeedfd9d1054e0071a2d50cc16a6c4659f681/src/hall.rs#L9) in your `temp_sensor` module. But instead of `HallSensor` of course use `TemperatureSensor` for the peripheral name.
- Then, you just need to add the `TemperatureSensor` as a `temperature_sensor` member in the `Peripherals` structure, and then change the `TemperatureSensorDriver::new` to take the peripheral, using the `impl Pripheral<P = TemperatureSensor> + 'd` magic you can copy from everywhere (this allows the driver to take the peripheral by move, or by `&mut`.

Also, what do you think about renaming `TemperatureSensor*` everywhere to just `TempSensor*`?
- `TemperatureSensor` is a bit too long
- Module is already named `temp_sensor`

(The peripheral instance in the `Peripherals` struct can then also be named `temp_sensor`.)

Reply:
> Also, what do you think about renaming TemperatureSensor* everywhere to just TempSensor*?

Done in 85f6a7cc15deed5822631d0b940e177bed38ad74